All_Relationships/6.true_relationship_preparation.py

```diff
@@ -0,0 +1,238 @@
+    cond_df = candidate_df.query("disease_id != @disease & gene_id == @gene")
+    b = cond_df["sentence_count"].sum() + 1
+    
+    cond_df = candidate_df.query("disease_id == @disease & gene_id != @gene")
+    c = cond_df["sentence_count"].sum() + 1
+    
+    cond_df = candidate_df.query("disease_id != @disease & gene_id != @gene")
+    d = cond_df["sentence_count"].sum() + 1
+    
+    c_table = np.array([[a, b], [c, d]])
+    
+    # Gather confidence interval
+    delta, ci, corrected_ci = diffprop(c_table)
+    lower_ci.append(ci[0])
+    
+    # Gather corrected odds ratio and p_values
+    odds_ratio, p_value = fisher_exact(c_table, alternative='greater')
+    odds.append(odds_ratio)
+    p_val.append(p_value)
+    
+    
+    total_disease = candidate_df[candidate_df["disease_id"] == disease]["sentence_count"].sum()
+    total_gene = candidate_df[candidate_df["gene_id"] == gene]["sentence_count"].sum()
+    expected.append((total_gene * total_disease)/float(total))
+
+candidate_df["nlog10_p_value"] = -np.log10(p_val)
+candidate_df["odds_ratio"] = odds
+candidate_df["expected_sen"] = expected
+candidate_df["lower_ci"] = lower_ci
+
+
+# In[ ]:
+
+
+candidate_df.sort_values("nlog10_p_value", ascending=False).head(20)
+
+
+# # Combine Sentence Marginal Probabilities
+
+# In this section we incorporate the marginal probabilites that are calculated from the bi-directional LSTM used in the [previous notebook](4.sentence-level-prediction.ipynb). For each sentence we grouped them by their disease-gene mention and report their marginal probabilites in different quantiles (0, 0.2, 0.4, 0.6, 0.8). Lastly we took the average of each sentence marginal to generate the "avg_marginal" column.
+
+# In[ ]:
+
+
+train_marginals_df = pd.read_csv("stratified_data/lstm_disease_gene_holdout/lstm_train_marginals.csv")
+dev_marginals_df = pd.read_csv("stratified_data/lstm_disease_gene_holdout/lstm_dev_marginals.csv")
+test_marginals_df = pd.read_csv("stratified_data/lstm_disease_gene_holdout/lstm_test_marginals.csv")
+
+
+# In[ ]:
+
+
+train_sentences_df = pd.read_csv("stratified_data/lstm_disease_gene_holdout/train_candidates_sentences.csv")
+dev_sentences_df = pd.read_csv("stratified_data/lstm_disease_gene_holdout/dev_candidates_sentences.csv")
+test_sentences_df = pd.read_csv("stratified_data/lstm_disease_gene_holdout/test_candidates_sentences.csv")
+
+
+# In[ ]:
+
+
+train_sentences_df["marginals"] = train_marginals_df["RNN_marginals"].values
+dev_sentences_df["marginals"] = dev_marginals_df["RNN_marginals"].values
+test_sentences_df["marginals"] = test_marginals_df["RNN_10_Marginals"].values
+
+
+# In[ ]:
+
+
+candidate_marginals = (
+    train_sentences_df[["disease_id", "gene_id", "marginals"]]
+    .append(dev_sentences_df[["disease_id", "gene_id", "marginals"]])
+    .append(test_sentences_df[["disease_id", "gene_id", "marginals"]])
+    )
+
+
+# In[ ]:
+
+
+group = candidate_marginals.groupby(["disease_id", "gene_id"])
```

Comment:
This is repetitive. The preferred solution is to define a list with the desired quantiles. This would also be more modular, allowing easy update of the selected quantiles.

You can automatically generate the column names like:

```python
'quantile_{:.2f}'.format(quantile)
```